### PR TITLE
Store cluster UUID in Lucene commits user data

### PR DESCRIPTION
**Pull request by fcofdez**
Relates ES-5412

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/gateway/ClusterStateUpdaters.java [hunk 1]
```diff
@@ -34,7 +34,10 @@ public static ClusterState setLocalNode(final ClusterState clusterState, Discove
             .build();
     }
 
-    static ClusterState upgradeAndArchiveUnknownOrInvalidSettings(final ClusterState clusterState, final ClusterSettings clusterSettings) {
+    public static ClusterState upgradeAndArchiveUnknownOrInvalidSettings(
```
Comment (fcofdez) [inline server/src/main/java/org/elasticsearch/gateway/ClusterStateUpdaters.java:37]: This is a bit unrelated, but we need access to this method in order to read the cluster state from the blob store.

**Review by DaveCTurner — COMMENTED** [review]
Makes sense to me, except that the cluster UUID might change if it's not committed (see `org.elasticsearch.cluster.metadata.Metadata#clusterUUIDCommitted`) so I think we should either only record _committed_ cluster UUIDs or else we should also record the committed flag.

Also I think the BwC test failure is genuine.

File: server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java [hunk 2]
```diff
@@ -602,19 +614,30 @@ public OnDiskState loadOnDiskState(Path dataPath, DirectoryReader reader) throws
...
+            onDiskStateMetadata.clusterUUID(),
+            builder.build()
+        );
+    }
+
+    public OnDiskStateMetadata readOnDiskStateMetadata(DirectoryReader reader) throws IOException {
         final Map<String, String> userData = reader.getIndexCommit().getUserData();
         logger.trace("loaded metadata [{}] from [{}]", userData, reader.directory());
-        assert userData.size() == COMMIT_DATA_SIZE : userData;
+        assert userData.size() == (userData.containsKey(CLUSTER_UUID_KEY) ? COMMIT_DATA_SIZE : COMMIT_DATA_SIZE - 1) : userData;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/PersistedClusterStateService.java:631]: Could we make assertion a function of the node version?

**Comment by fcofdez** [comment]
Looks like `PersistedClusterStateService#overrideVersion` should try to read back the cluster state if we're forcing the version to be >= 8.8 in order to populate the cluster uuid fields

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
@DaveCTurner could you review this PR when you have a chance? Thanks!

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
Thanks David!

**Pull request merged**